insights: nil pointer in lockingRegistry.ObserveTransaction

When running TPCH query 18 with concurrency 512 using a custom [branch](https://github.com/yuzefovich/cockroach/tree/gomemlimit) on 3 node cluster with `n1-standard-4` machines while having set `GOMEMLIMIT=7.5GiB`, I hit the following on two out of three nodes after a few minutes of load:
```
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1f855f8]

goroutine 352 [running]:
panic({0x4c3c160, 0x964cd20})
	GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc0079919b8 sp=0xc0079918f8 pc=0x48ab3a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0000ce6c0, {0x68a73e8, 0xc0033fb890})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:248 +0xa5 fp=0xc007991a00 sp=0xc0079919b8 pc=0x1043d85
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2.3()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x2e fp=0xc007991a28 sp=0xc007991a00 pc=0x104514e
runtime.deferCallSave(0xc007991af8, 0xc0038def90?)
	GOROOT/src/runtime/panic.go:796 +0x88 fp=0xc007991a38 sp=0xc007991a28 pc=0x48a728
runtime.runOpenDeferFrame(0xc093a95630?, 0xc010f92c80)
	GOROOT/src/runtime/panic.go:769 +0x1a5 fp=0xc007991a80 sp=0xc007991a38 pc=0x48a545
panic({0x4c3c160, 0x964cd20})
	GOROOT/src/runtime/panic.go:884 +0x212 fp=0xc007991b40 sp=0xc007991a80 pc=0x48a992
runtime.panicmem(...)
	GOROOT/src/runtime/panic.go:260
runtime.sigpanic()
	GOROOT/src/runtime/signal_unix.go:835 +0x2f6 fp=0xc007991b90 sp=0xc007991b40 pc=0x4a19d6
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights.(*statementBuf).release(0xfffffffe?)
	github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights/pkg/sql/sqlstats/insights/registry.go:52 +0x18 fp=0xc007991bb8 sp=0xc007991b90 pc=0x1f855f8
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights.(*lockingRegistry).ObserveTransaction.func1()
	github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights/pkg/sql/sqlstats/insights/registry.go:71 +0x26 fp=0xc007991bd0 sp=0xc007991bb8 pc=0x1f85b66
runtime.deferCallSave(0xc007991ca0, 0xc0038dedc8?)
	GOROOT/src/runtime/panic.go:796 +0x88 fp=0xc007991be0 sp=0xc007991bd0 pc=0x48a728
runtime.runOpenDeferFrame(0x686b680?, 0xc093a95630)
	GOROOT/src/runtime/panic.go:769 +0x1a5 fp=0xc007991c28 sp=0xc007991be0 pc=0x48a545
panic({0x4c3c160, 0x964cd20})
	GOROOT/src/runtime/panic.go:884 +0x212 fp=0xc007991ce8 sp=0xc007991c28 pc=0x48a992
runtime.panicmem(...)
	GOROOT/src/runtime/panic.go:260
runtime.sigpanic()
	GOROOT/src/runtime/signal_unix.go:835 +0x2f6 fp=0xc007991d38 sp=0xc007991ce8 pc=0x4a19d6
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights.(*lockingRegistry).ObserveTransaction(0xc002951650, {{0xc06d5ed0c0?, 0xc0038deedc?}}, 0xc01289dc80)
	github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights/pkg/sql/sqlstats/insights/registry.go:74 +0x124 fp=0xc007991e08 sp=0xc007991d38 pc=0x1f85804
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights.(*concurrentBufferIngester).ingest(0xc002951680, 0xc07f1f0000)
	github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights/pkg/sql/sqlstats/insights/ingester.go:108 +0x107 fp=0xc007991ea8 sp=0xc007991e08 pc=0x1f844c7
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights.(*concurrentBufferIngester).Start.func1({0xc003720000?, 0xc003231e30?})
	github.com/cockroachdb/cockroach/pkg/sql/sqlstats/insights/pkg/sql/sqlstats/insights/ingester.go:72 +0x4a fp=0xc007991f30 sp=0xc007991ea8 pc=0x1f8430a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:489 +0x146 fp=0xc007991fe0 sp=0xc007991f30 pc=0x1045006
runtime.goexit()
	GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc007991fe8 sp=0xc007991fe0 pc=0x4bf841
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:480 +0x43b
```

Note that this is a complete overload scenario where CPU is pegged at 100%, so most queries error out. Many goroutines wait a long time before they get scheduled onto the CPUs. Still, there appears to be some synchronization missing around insights handling (I think).

[stack.zip](https://github.com/cockroachdb/cockroach/files/9995581/stack.zip)



Jira issue: CRDB-21433

--- a/pkg/sql/sqlstats/insights/registry.go
+++ b/pkg/sql/sqlstats/insights/registry.go
@@ -66,7 +66,10 @@ func (r *lockingRegistry) ObserveTransaction(sessionID clusterunique.ID, transac
 	if !r.enabled() {
 		return
 	}
-	statements := r.statements[sessionID]
+	statements, ok := r.statements[sessionID]
+	if !ok {
+		return
+	}
 	delete(r.statements, sessionID)
 	defer statements.release()
 

--- a/pkg/sql/sqlstats/insights/registry_test.go
+++ b/pkg/sql/sqlstats/insights/registry_test.go
@@ -218,4 +218,10 @@ func TestRegistry(t *testing.T) {
 
 		require.Equal(t, expected, actual)
 	})
+
+	t.Run("txn with no stmts", func(t *testing.T) {
+		st := cluster.MakeTestingClusterSettings()
+		registry := newRegistry(st, &latencyThresholdDetector{st: st}, newStore(st))
+		require.NotPanics(t, func() { registry.ObserveTransaction(session.ID, transaction) })
+	})
 }

